File: ecommerce/src/main/java/com/bookstore/ecommerce/app/domain/data/Customer.java
package com.bookstore.ecommerce.app.domain.data;

import org.modelmapper.ModelMapper;
import lombok.Builder;
import lombok.Data;

@Builder
@Data
public class Customer {
  private String id;
  private String name;
  private String phone;
  private String deliveryAddress;

  public Customer clone() {
    return new ModelMapper().map(this, Customer.class);
  }
}

File: order/app/domain/customer.go
package domain

type CustomerData struct {
	Id              string
	Name            string
	Phone           string
	DeliveryAddress string
}

func (c *CustomerData) Clone() CustomerData {
	return CustomerData{Id: c.Id, Name: c.Name, Phone: c.Phone, DeliveryAddress: c.DeliveryAddress}
}

type Customer struct {
	customer CustomerData
}

func (Customer) New(customer CustomerData) *Customer {
	cloned := customer.Clone()
	if cloned.Id == "" {
		cloned.Id = NewId()
	}

	return &Customer{customer: cloned}
}

func (customer *Customer) State() CustomerData {
	return customer.customer.Clone()
}
